### PR TITLE
[react] add 'this : void' to event handlers

### PR DESCRIPTION
**Pull request by hexagonal-sun**
Add the 'this : void' this-parameter (see [1]) to event handlers.

[1]: https://www.typescriptlang.org/docs/handbook/functions.html#this-parameters-in-callbacks

Please fill in this template.

- [ x ] Use a meaningful title for the pull request. Include the name of the package modified.
- [ x ] Test the change in your own code. (Compile and run.)
- [ x ] Add or edit tests to reflect the change. (Run with `npm test`.)
- [ x ] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [ x ] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [ x ] Provide a URL to documentation or source code which provides context for the suggested changes: https://www.typescriptlang.org/docs/handbook/functions.html#this-parameters-in-callbacks

### CODE REVIEW

**Review by ferdaber — APPROVED** [review]

**Review by eps1lon — CHANGES_REQUESTED** [review]
Looks good. However, we should add a test for this change.

**Comment by ExE-Boss** [comment]
Note that the **dtslint** [`void‑return` rule](https://github.com/microsoft/dtslint/blob/master/docs/void-return.md) currently disallows `this: void`

File: types/react/index.d.ts [hunk 1]
```diff
@@ -1284,7 +1284,7 @@ declare namespace React {
     // Event Handler Types
     // ----------------------------------------------------------------------
 
-    type EventHandler<E extends SyntheticEvent<any>> = { bivarianceHack(event: E): void }["bivarianceHack"];
+    type EventHandler<E extends SyntheticEvent<any>> = { bivarianceHack(this : void, event: E): void }["bivarianceHack"];
```
Comment (ExE-Boss) [inline types/react/index.d.ts:1287]: <https://github.com/DefinitelyTyped/DefinitelyTyped/pull/45834/checks#step:5:1917>

```suggestion
    type EventHandler<E extends SyntheticEvent<any>> = { bivarianceHack(this: void, event: E): void }["bivarianceHack"];
```

**Review by ExE-Boss — COMMENTED** [review]
The test failure at <https://github.com/DefinitelyTyped/DefinitelyTyped/pull/45834/checks#step:5:528> seems related to <https://github.com/microsoft/dtslint/issues/278>, which is a duplicate of <https://github.com/microsoft/dtslint/issues/281> and <https://github.com/microsoft/dtslint/issues/293>.

---

Note that the **dtslint** [`void‑return` rule](https://github.com/microsoft/dtslint/blob/master/docs/void-return.md) currently disallows `this: void`.

**Comment by hexagonal-sun** [comment]
> Looks good. However, we should add a test for this change.

@eps1lon Thanks for taking a look.  I was wondering what kind of test I could add; I'm not sure how to make the test framework check that an error would be raised with a specific test. 

One could cause an error with:

```typescript
class SyntheticEventTargetValue extends React.Component<{}, { value: string }> {
    state: { value: string };
    constructor(props: {}) {
        super(props);
        this.state = { value: 'a' };
    }

    private callback(this: this, e: React.ChangeEvent<HTMLTextAreaElement>)
    {
    }

    render(this : this) {
        return DOM.textarea({
            value: this.state.value,
            onChange: this.callback,
        });
    }
}

```
In which case you get:
```
No overload matches this call.
  Overload 1 of 2, '(props?: DetailedHTMLProps<TextareaHTMLAttributes<HTMLTextAreaElement>, HTMLTextAreaElement> | null | undefined, ...children: ReactNode[]): DetailedReactHTMLElement<...>', gave the following error.
    Type '(this: this, e: ChangeEvent<HTMLTextAreaElement>) => void' is not assignable to type '(this: void, event: ChangeEvent<HTMLTextAreaElement>) => void'.
      The 'this' types of each signature are incompatible.
        Type 'void' is not assignable to type 'this'.
          'this' could be instantiated with an arbitrary type which could be unrelated to 'void'.
  Overload 2 of 2, '(props?: DetailedHTMLProps<TextareaHTMLAttributes<HTMLTextAreaElement>, HTMLTextAreaElement> | null | undefined, ...children: ReactNode[]): DOMElement<...>', gave the following error.
    Type '(this: this, e: ChangeEvent<HTMLTextAreaElement>) => void' is not assignable to type '(this: void, event: ChangeEvent<HTMLTextAreaElement>) => void'.
```
Other than that, I suppose the fact that it hasn't broke other event handler tests is a good thing.

**Comment by eps1lon** [comment]
> One could cause an error with:

That seems like the perfect test case for this PR. Not binding event handlers is a common mistake and having a type check for it would help a lot. I'd simplify it a bit to 

```tsx
class TextArea extends React.Component<{}, { value: string }> {
    state: { value: string };
    constructor(props: {}) {
        super(props);
        this.state = { value: 'a' };
    }

    handleChange(e: React.ChangeEvent<HTMLTextAreaElement>) {
        this.setState({ value: e.target.value })
    }

    render() {
        return <textarea onChange={this.handleChange} />
    }
}
```
which gets rid of some noise (unnecessary `this`, `private` keyword).

That component would pass type checking but fail at runtime.

**Comment by hexagonal-sun** [comment]
> That seems like the perfect test case for this PR

Agreed, however your case as it stands woudln't catch the problem.  You'd need to change the line

```typescript
    handleChange(e: React.ChangeEvent<HTMLTextAreaElement>) {
```
to
```typescript
    handleChange(this: this, e: React.ChangeEvent<HTMLTextAreaElement>) {
```

If was was to commit this, then that would cause an (expected) error when running the tests.  Is there a way I can instruct the test framework to require an error with a test case?

**Comment by eps1lon** [comment]
> You'd need to change the line

Why? Typescript already infers `this` as far as I can tell.

**Comment by eps1lon** [comment]
>  Is there a way I can instruct the test framework to require an error with a test case?

`$ExpectError` or `$ExpectType`

**Comment by hexagonal-sun** [comment]
> Why? Typescript already infers `this` as far as I can tell.

I don't think it does.  The following example doesn't cause an error, when I think it should:

```typescript
function setupCallback(a : (this : void) => void) : void
{
    a();
}

class foobar
{
    memberFunc()
    {
    }

    doError(this : this)
    {
        setupCallback(this.memberFunc)
    }
}
```

However, if I change the definition of `memberFunc` to:

```typescript
    memberFunc(this: this)
```

Then I get the error:

```
Argument of type '(this: this) => void' is not assignable to parameter of type '(this: void) => void'.
  The 'this' types of each signature are incompatible.
    Type 'void' is not assignable to type 'this'.
      'this' could be instantiated with an arbitrary type which could be unrelated to 'void'.
```

I don't understand why `this` isn't implicitly typed.

**Comment by sandersn** [comment]
Typescript does not infer `this: this` for methods, even though you would expect it to. I wrote this feature, but it's been a long time, so I dug up some discussion on it. As far as I can tell, the concern is mostly about breaking existing code, then consistency. Adding `this: this` to methods probably wouldn't break much if no other changes are made, but it's inconsistent in two directions:

1. Why not make `this: void` (or `this: undefined`) in functions? Well, it would break way more _admittedly suspicious code_.
2. Why not make `this: this` for methods on interfaces? This is important for heavy OO codebases which use variable with interface types to point to class instances. (eg `var l: IList = new List()`). But this breaks tonnes of probably-fine code, and causes compile times to get worse because all interfaces become generic.

It's possible that the consistency argument is a paper tiger, so it might be worth revisiting this issue. For now, though, this PR only provides a benefit to people who take the time to explicitly annotate `this: this` in methods. Is it still worthwhile? The compile-time performance looks better, but it could be a fluke.

Background:
* Original PR, with discussion near the bottom: https://github.com/microsoft/TypeScript/pull/6739
* Issue tracking this feature: https://github.com/microsoft/TypeScript/issues/7968
* Design meeting notes: https://github.com/microsoft/TypeScript/issues/7689

**Comment by hexagonal-sun** [comment]
@sandersn Many thanks for the info, that has helped clear things up.

It looks as though this change is only really useful for people that annotate member functions with `this : this`. Whether this change is worth it is something I'm happy to leave to the maintainers of this type package.

Thoughts?

**Comment by RyanCavanaugh** [comment]
I'd really prefer to see a test here so we could weigh the effects of `: void` vs `; unknown`. Untested changes in probably the second most popular package on DT aren't really a good idea.

**Comment by ExE-Boss** [comment]
Or `this: undefined`, which matches what happens when a function is called without a `this` context in strict mode.

**Review by RyanCavanaugh — CHANGES_REQUESTED** [review]
Please add tests demonstrating what impact this is intended to have

**Pull request closed without merging**